src/vm/costs/mod.rs

```diff
@@ -194,6 +260,235 @@ pub enum CostErrors {
+        Some(serialized) => serde_json::from_str(&serialized).unwrap(),
+        None => return Ok(CostStateSummary::empty()),
+    };
+    Ok(CostStateSummary::from(serialized))
+}
+
+fn store_state_summary(
+    clarity_db: &mut ClarityDatabase,
+    to_store: &CostStateSummary,
+) -> Result<()> {
+    let block_height = clarity_db.get_current_block_height();
+
+    clarity_db.put(
+        "vm-costs::last-processed-at-height",
+        &Value::UInt(block_height as u128),
+    );
+    let serialized_summary =
+        serde_json::to_string(&SerializedCostStateSummary::from(to_store.clone()))
+            .expect("BUG: failure to serialize cost state summary struct");
+    clarity_db.set_metadata(
+        &STACKS_BOOT_COST_VOTE_CONTRACT,
+        "state_summary",
+        &serialized_summary,
+    );
+
+    Ok(())
+}
+
+///
+/// This method loads a cost state summary structure from the currently open stacks chain tip
+///   In doing so, it reads from the cost-voting contract to find any newly confirmed proposals,
+///    checks those proposals for validity, and then applies those changes to the cached set
+///    of cost functions.
+///
+fn load_cost_functions(clarity_db: &mut ClarityDatabase) -> Result<CostStateSummary> {
+    let last_processed_count = clarity_db
+        .get_value("vm-costs::last_processed_count", &TypeSignature::UIntType)
+        .unwrap_or(Value::UInt(0))
+        .expect_u128();
+    let cost_voting_contract = &STACKS_BOOT_COST_VOTE_CONTRACT;
+    let confirmed_proposals_count = clarity_db
+        .lookup_variable(&cost_voting_contract, "confirmed-proposal-count")
+        .map_err(|e| CostErrors::CostComputationFailed(e.to_string()))?
+        .expect_u128();
+    debug!("Check cost voting contract";
+           "confirmed_proposal_count" => confirmed_proposals_count,
+           "last_processed_count" => last_processed_count);
+
+    // we need to process any confirmed proposals in the range [fetch-start, fetch-end)
+    let (fetch_start, fetch_end) = (last_processed_count, confirmed_proposals_count);
+    let mut state_summary = load_state_summary(clarity_db)?;
+    for confirmed_proposal in fetch_start..fetch_end {
+        // fetch the proposal data
+        let entry = clarity_db
+            .fetch_entry(
+                &cost_voting_contract,
+                "confirmed-proposals",
+                &Value::from(
+                    TupleData::from_data(vec![(
+                        "confirmed-id".into(),
+                        Value::UInt(confirmed_proposal),
+                    )])
+                    .unwrap(),
+                ),
+            )
+            .expect("BUG: Failed querying confirmed-proposals")
+            .expect_optional()
+            .expect("BUG: confirmed-proposal-count exceeds stored proposals")
+            .expect_tuple();
+        let target_contract = match entry
+            .get("function-contract")
+            .unwrap()
```

Comment:
Not a bug, but maybe use an `expect()` instead of an `unwrap()`?

Reply:
Addressed in ee998d20dc043c9cac68669aa8f03a7a8a02f407